traits/ctraits.c

```diff
@@ -310,9 +310,27 @@ fatal_trait_error ( void ) {
 +----------------------------------------------------------------------------*/
 
 static int
-invalid_attribute_error ( void ) {
+invalid_attribute_error ( PyObject * name ) {
 
-    PyErr_SetString( PyExc_TypeError, "attribute name must be string" );
+    // Build error message by collecting the type of the object received
+    PyTypeObject* obj_type = name->ob_type;
+    const char* type_name = obj_type->tp_name;
+    PyObject * ob_repr = PyObject_Repr(name);
+
+#if PY_MAJOR_VERSION >= 3
+
+    const char* fmt = "attribute name must be string. Got %R (%.200s).";
+
+    PyErr_Format( PyExc_TypeError, fmt, ob_repr, type_name );
+
+#else
+
+    const char* obj_repr_str = PyString_AsString(ob_repr);
+    const char* fmt = "attribute name must be string. Got %.200s (%.200s).";
```

Comment:
Since `unicode` objects might reasonably be considered to be strings, what do you think about a more specific error message here?  E.g. "attribute name must be an instance of <type 'str'>.  Got ...".